things/policies/service.go

```diff
@@ -129,12 +140,45 @@ func (svc service) AddPolicy(ctx context.Context, token string, p Policy) (Polic
 	// If the client has `g_add` action on the object or is the owner of the object, add the policy
 	ar := AccessRequest{Subject: userID, Object: p.Object, Action: "g_add"}
 	if _, err := svc.policies.EvaluateGroupAccess(ctx, ar); err == nil {
+		if err := svc.checkSubject(ctx, userID, external, p); err != nil {
+			return Policy{}, err
+		}
+
 		return svc.policies.Save(ctx, p)
 	}
 
 	return Policy{}, errors.ErrAuthorization
 }
 
+// checkSubject is used to check if the subject is valid.
+//
+// 1. If the subject is a thing (external is false), check the following:
+//   - The user is the owner of the thing
+//   - The user has `c_share` action on the thing
+//
+// 2. If the subject is a user (external is true), check the following:
+//   - The user is the owner of the user
+//   - The user has `c_share` action on the user
+func (svc service) checkSubject(ctx context.Context, userID string, external bool, p Policy) error {
+	switch external {
+	case false:
+		ar := AccessRequest{Subject: userID, Object: p.Subject, Action: sharePolicyAction}
+		if _, err := svc.policies.EvaluateThingAccess(ctx, ar); err == nil {
+			return nil
+		}
+
+		return errors.ErrAuthorization
+	case true:
+		if err := svc.usersAuthorize(ctx, userID, p.Subject, sharePolicyAction, ClientEntityType); err == nil {
+			return nil
+		}
+
+		return errors.ErrAuthorization
+	default:
+		return ErrInvalidClient
+	}
```

Comment:
I let you address what is wrong with this. 🙂

Reply:
This is still not addressed. We have three cases for the bool variable, which makes no sense.